internal/graph/graph_test.go

```diff
@@ -0,0 +1,681 @@
+						},
+					},
+					{
+						Type: "folder",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": typesystem.This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("user", ""),
+										typesystem.RelationReference("group", "member"),
+									},
+								},
+							},
+						},
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"parent": typesystem.This(),
+							"viewer": typesystem.TupleToUserset("parent", "viewer"),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"parent": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("folder", ""),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			target: typesystem.RelationReference("document", "viewer"),
+			source: typesystem.RelationReference("folder", "viewer"),
+			expected: []*RelationshipIngress{
+				{
+					Type:             TupleToUsersetIngress,
+					Ingress:          typesystem.RelationReference("document", "viewer"),
+					TuplesetRelation: typesystem.RelationReference("document", "parent"),
+				},
+			},
+		},
+		{
+			name: "Undefined relation on one type involved in a tuple to userset",
+			model: &openfgapb.AuthorizationModel{
+				TypeDefinitions: []*openfgapb.TypeDefinition{
+					{
+						Type: "user",
+					},
+					{
+						Type: "organization",
+					},
+					{
+						Type: "folder",
+						Relations: map[string]*openfgapb.Userset{
+							"viewer": typesystem.This(),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"viewer": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("user", ""),
+									},
+								},
+							},
+						},
+					},
+					{
+						Type: "document",
+						Relations: map[string]*openfgapb.Userset{
+							"parent": typesystem.This(),
+							"viewer": typesystem.TupleToUserset("parent", "viewer"),
+						},
+						Metadata: &openfgapb.Metadata{
+							Relations: map[string]*openfgapb.RelationMetadata{
+								"parent": {
+									DirectlyRelatedUserTypes: []*openfgapb.RelationReference{
+										typesystem.RelationReference("folder", ""),
+										typesystem.RelationReference("organization", ""),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			target: typesystem.RelationReference("document", "viewer"),
+			source: typesystem.RelationReference("user", ""),
+			expected: []*RelationshipIngress{
+				{
+					Type:    DirectIngress,
+					Ingress: typesystem.RelationReference("folder", "viewer"),
+				},
+			},
+		},
+		{
+			name: "Nested group membership returns only top-level ingress",
```

Comment:
@craigpastro I added this test just now. We're currently failing this test.

```
type user

type group
  relations
    define member: [user, group#member]
```

`RelationshipIngresses(group#member, user)` currently returns `[group#member, group#member]`, but it should only return one occurrence of `group#member`.

Reply:
Squashed. Thanks!